File: modules/indigo-connectivity/android/src/main/java/dev/daniell/indigo/modules/connectivity/IndigoConnectivityModule.kt
package dev.daniell.indigo.modules.connectivity

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.launch

data class ConnectivityState(val isConnected: Boolean, val type: String)

// Función de nivel de paquete (no un método privado) a propósito, igual que en los
// módulos de las FASES 6-9: toma booleanos primitivos, no NetworkCapabilities (una
// clase del framework de Android que no se puede instanciar/mockear en un test JUnit
// puro) — así es 100% testeable sin Robolectric ni un dispositivo real.
internal fun connectivityStateFrom(
  hasInternet: Boolean,
  hasWifi: Boolean,
  hasCellular: Boolean,
): ConnectivityState {
  if (!hasInternet) return ConnectivityState(isConnected = false, type = "none")

  val type = when {
    hasWifi -> "wifi"
    hasCellular -> "cellular"
    else -> "unknown"
  }
  return ConnectivityState(isConnected = true, type = type)
}

private fun connectivityStateFromCapabilities(capabilities: NetworkCapabilities?): ConnectivityState {
  if (capabilities == null) return ConnectivityState(isConnected = false, type = "none")

  return connectivityStateFrom(
    hasInternet = capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET),
    hasWifi = capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI),
    hasCellular = capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR),
  )
}

private fun ConnectivityState.toMap(): Map<String, Any?> =
  mapOf("isConnected" to isConnected, "type" to type)

// Coroutines + Flow: envuelve la API de ConnectivityManager, basada en callbacks
// (NetworkCallback), en un callbackFlow — el patrón estándar de Kotlin para exponer
// una API de callbacks como un Flow frío y cancelable. awaitClose garantiza que
// unregisterNetworkCallback se llama siempre que el Flow se cancele (cuando JS deja de
// escuchar el evento), sin importar cómo termine la colección.
private fun observeConnectivity(context: Context): Flow<ConnectivityState> = callbackFlow {
  val connectivityManager =
    context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
  val request = NetworkRequest.Builder()
    .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
    .build()

  val callback = object : ConnectivityManager.NetworkCallback() {
    override fun onCapabilitiesChanged(network: Network, capabilities: NetworkCapabilities) {
      trySend(connectivityStateFromCapabilities(capabilities))
    }

    override fun onLost(network: Network) {
      trySend(ConnectivityState(isConnected = false, type = "none"))
    }
  }

  connectivityManager.registerNetworkCallback(request, callback)
  awaitClose { connectivityManager.unregisterNetworkCallback(callback) }
}.distinctUntilChanged()

class IndigoConnectivityModule : Module() {
  private var observationJob: Job? = null

  override fun definition() = ModuleDefinition {
    Name("IndigoConnectivity")
    Events("onConnectivityChange")

    AsyncFunction("getCurrentState") {
      currentConnectivityState()
    }

    // Se suscribe al Flow solo cuando JS tiene al menos un listener activo — evita
    // tener un NetworkCallback registrado (y consumiendo batería) sin que nadie del
    // lado JS esté escuchando.
    OnStartObserving("onConnectivityChange") {
      val context = appContext.reactContext ?: return@OnStartObserving
      observationJob = appContext.modulesQueue.launch {
        observeConnectivity(context).collect { state ->
          sendEvent("onConnectivityChange", state.toMap())
        }
      }
    }

    OnStopObserving("onConnectivityChange") {
      observationJob?.cancel()
      observationJob = null
    }
  }

  private fun currentConnectivityState(): Map<String, Any?> {
    val context = appContext.reactContext
      ?: return ConnectivityState(isConnected = false, type = "none").toMap()

    val connectivityManager =
      context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    val capabilities = connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
    return connectivityStateFromCapabilities(capabilities).toMap()
  }
}

File: modules/indigo-connectivity/ios/IndigoConnectivityModule.swift
import ExpoModulesCore
import Network

struct ConnectivityState {
  let isConnected: Bool
  let type: String
}

// Función libre (no un método privado), igual que en los módulos de las FASES 6-9:
// toma booleanos primitivos, no NWPath (una clase del framework que no se puede
// instanciar en un test XCTest puro) — así es 100% testeable sin un monitor real.
func connectivityState(isConnected: Bool, usesWifi: Bool, usesCellular: Bool) -> ConnectivityState {
  guard isConnected else {
    return ConnectivityState(isConnected: false, type: "none")
  }

  let type: String
  if usesWifi {
    type = "wifi"
  } else if usesCellular {
    type = "cellular"
  } else {
    type = "unknown"
  }
  return ConnectivityState(isConnected: true, type: type)
}

private func connectivityState(from path: NWPath) -> ConnectivityState {
  connectivityState(
    isConnected: path.status == .satisfied,
    usesWifi: path.usesInterfaceType(.wifi),
    usesCellular: path.usesInterfaceType(.cellular)
  )
}

private func connectivityStateDict(_ state: ConnectivityState) -> [String: Any?] {
  ["isConnected": state.isConnected, "type": state.type]
}

// AsyncStream: envuelve el pathUpdateHandler basado en callback de NWPathMonitor — el
// equivalente Swift exacto de callbackFlow en Kotlin. onTermination garantiza que el
// monitor se cancela siempre que el stream se cancele (cuando JS deja de escuchar el
// evento), sin importar cómo termine el consumo.
private func connectivityUpdates() -> AsyncStream<ConnectivityState> {
  AsyncStream { continuation in
    let monitor = NWPathMonitor()
    let queue = DispatchQueue(label: "dev.daniell.indigo.modules.connectivity")

    monitor.pathUpdateHandler = { path in
      continuation.yield(connectivityState(from: path))
    }
    continuation.onTermination = { _ in
      monitor.cancel()
    }

    monitor.start(queue: queue)
  }
}

public class IndigoConnectivityModule: Module {
  private var observationTask: Task<Void, Never>?

  public func definition() -> ModuleDefinition {
    Name("IndigoConnectivity")
    Events("onConnectivityChange")

    AsyncFunction("getCurrentState") { () -> [String: Any?] in
      await currentConnectivityState()
    }

    // Se suscribe al AsyncStream solo cuando JS tiene al menos un listener activo —
    // evita tener un NWPathMonitor corriendo sin que nadie del lado JS esté escuchando.
    OnStartObserving("onConnectivityChange") {
      observationTask = Task {
        for await state in connectivityUpdates() {
          sendEvent("onConnectivityChange", connectivityStateDict(state))
        }
      }
    }

    OnStopObserving("onConnectivityChange") {
      observationTask?.cancel()
      observationTask = nil
    }
  }

  private func currentConnectivityState() async -> [String: Any?] {
    await withCheckedContinuation { continuation in
      let monitor = NWPathMonitor()
      let queue = DispatchQueue(label: "dev.daniell.indigo.modules.connectivity.snapshot")

      monitor.pathUpdateHandler = { path in
        let state = connectivityState(from: path)
        monitor.cancel()
        continuation.resume(returning: connectivityStateDict(state))
      }
      monitor.start(queue: queue)
    }
  }
}
